core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,210 @@
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.io.Closeable;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.ScheduledExecutorService;
+
+import javax.annotation.Nullable;
+
+
+/**
+ * --Add warning about using this too liberally--
+ * --Add warning about using this with containers that unload Classloaders--
+ *
+ *
+ */
+// Discussion point. Using java.io.Closeable here so we can inherit AutoCloseable where its available.
+public abstract class Context implements Closeable {
+
+  /**
+   * The logical root context which is 'current' if no other context is bound.
+   */
+  public static final Context ROOT = null; // FIXME
+
+  public interface Listener {
+    public void closed(Context context);
+  }
+
+  /**
+   * Create a {@link Key} with the given name.
+   */
+  public static <T> Key<T> key(String name) {
+    return new Key<T>(name);
+  }
+
+  /**
+   * Create a {@link Key} with the given name and default value.
+   */
+  public static <T> Key<T> keyWithDefault(String name, T defaultValue) {
```

Comment:
It allows for defining defaults at the definition site which would have clarity benefits over creating a 'base' context and installing 'defaults' into it in some frameworky code.